Handling resetting of PushToApplicationPreferences

Hamza-Azeem opened the pull request:
### **User description**
Closes #14410

Handled PushToApplicationPreferences removed default config from JabRefCliPreferences, added private constructor for 
PushToApplicationPreferences to handle default config with getDefault method, setAll method and followed the rest of instructions. Moved logic of getEmptyIsDefault method from JabRefCliPreferences to PushToApplicationPreferences but didnt delete it from JabRefCliPreferences. Modified PushApplications class by adding a new instance variable(key).

### Steps to test

<!-- Describe how reviewers can test this fix/feature. Ideally, think of how you would guide a beginner user of JabRef to try out your change. -->
After running the application click on File then preferences, navigate to External programs.
Try changing Application to push  entries to option, restart the application try resetting preferences.

### Mandatory checks
<!--
Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [/] I added JUnit tests for changes (if applicable)
- [/] I added screenshots in the PR description (if change is visible to the user)
- [ ] I described the change in `CHANGELOG.md` in a way that is understandable for the average user (if change is visible to the user)
- [x] I checked the [user documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request updating file(s) in <https://github.com/JabRef/user-documentation/tree/main/en>.


___

### **PR Type**
Enhancement, Bug fix


___

### **Description**
- Moved PushToApplicationPreferences default configuration from JabRefCliPreferences to PushToApplicationPreferences class

- Added private constructor and getDefault() method for proper default initialization

- Implemented setAll() method to enable resetting preferences to defaults

- Updated getEmptyIsDefault() method signature to accept default value parameter

- Added preference reset logic in clear() and importPreferences() methods


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["JabRefCliPreferences"] -->|removed defaults| B["PushToApplicationPreferences"]
  B -->|added private constructor| C["getDefault()"]
  B -->|added method| D["setAll()"]
  E["clear() method"] -->|calls| D
  F["importPreferences() method"] -->|calls| D
```



<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>JabRefCliPreferences.java</strong><dd><code>Refactored PushToApplicationPreferences initialization and reset</code></dd></summary>
<hr>

jablib/src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java

<ul><li>Removed PushToApplicationPreferences default configuration block from <br>constructor<br> <li> Added getPushToApplicationPreferencesFromBackingStore() method to load <br>preferences from backing store with defaults<br> <li> Modified getPushToApplicationPreferences() to use new method and <br>PushToApplicationPreferences.getDefault()<br> <li> Updated getEmptyIsDefault() signature to accept defaultValue parameter<br> <li> Added setAll() calls in clear() and importPreferences() methods for <br>preference reset</ul>


</details>


  </td>
  <td><a href="https://github.com/JabRef/jabref/pull/14691/files#diff-2ab744cd10eeea9eccf84f1a10b40b79862bcd79ddb96775f8824d370e705b32">+34/-46</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>PushToApplicationPreferences.java</strong><dd><code>Added default initialization and reset functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

jablib/src/main/java/org/jabref/logic/push/PushToApplicationPreferences.java

<ul><li>Added private constructor with default configuration initialization<br> <li> Added getDefault() static method to create default preferences <br>instance<br> <li> Added setAll() method to reset all preference properties to another <br>instance's values<br> <li> Added imports for File, HashMap, and OS classes</ul>


</details>


  </td>
  <td><a href="https://github.com/JabRef/jabref/pull/14691/files#diff-f73a5377de382a2dbebb92042989f8933a6d0d68392762e61d96e0d2ec999761">+44/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Formatting</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>PushApplications.java</strong><dd><code>Minor formatting adjustments</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

jablib/src/main/java/org/jabref/logic/push/PushApplications.java

- Added blank lines for formatting consistency


</details>


  </td>
  <td><a href="https://github.com/JabRef/jabref/pull/14691/files#diff-cb0a9e570cf57310cc36d190a28bb684eb8826a8010ad19ed29e5c950def9aba">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

jabref-machine commented:
Your code currently does not meet JabRef's code guidelines. IntelliJ auto format covers some cases. There seem to be issues with your code style and autoformat configuration. Please reformat your code (<kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>L</kbd>) and commit, then push.

In special cases, consider using `// formatter:off` and `// formatter:on` annotations to allow deviation from the code style.


<!-- ghprcomment
Source Code Tests
format
-->

Hamza-Azeem commented:
Hello @calixtus, 

I have removed getEmptyIsDefault from PushToApplicationPreference class, added the static keys instead of getting them from PushApplications Enum and got PushApplications into it's initial state without the modifications I did in the last commit.
Please tell me if there is anything else to do.

koppor commented:
> Hello @calixtus,
> 
> I have removed getEmptyIsDefault from PushToApplicationPreference class, added the static keys instead of getting them from PushApplications Enum and got PushApplications into it's initial state without the modifications I did in the last commit. Please tell me if there is anything else to do.

Maybe check the CI output?

This is what you can address without someone telling you the exact steps

<img width="1085" height="631" alt="image" src="https://github.com/user-attachments/assets/ef0d0f6b-4ce7-4af5-a912-5c5b0406b5fb" />

<img width="1933" height="924" alt="image" src="https://github.com/user-attachments/assets/5c7470d7-43b3-4a32-86bf-4ba90d813b91" />


This is not respecting our time.

calixtus commented on jablib/src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java:
```diff
@@ -812,32 +791,40 @@ public void setLanguageDependentDefaultValues() {
+            String oldKey = lookup.get(entry.getKey());
+            String defaultValue =  entry.getValue();
+            entry.setValue(getEmptyIsDefault(oldKey, defaultValue));
+        }
```
This looks incredibly complicated for a very simple task.

calixtus commented on jablib/src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java:
```diff
@@ -926,8 +913,7 @@ public String get(String key) {
 
-    public String getEmptyIsDefault(String key) {
-        String defaultValue = (String) defaults.get(key);
+    public String getEmptyIsDefault(String key, String defaultValue) {
```
Should not be modified.

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushApplications.java:
```diff
@@ -18,11 +18,13 @@ public enum PushApplications {
     private final String id;
     private final String displayName;
 
+
```
checkstyle

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushApplications.java:
```diff
@@ -18,11 +18,13 @@ public enum PushApplications {
         this.displayName = displayName;
     }
 
+
```
checkstyle

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushApplications.java:
```diff
@@ -48,4 +50,6 @@ public String getId() {
         return displayName;
     }
+
+
```
checkstyle

calixtus reviewed (changes requested):
checkstyle

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushToApplicationPreferences.java:
```diff
@@ -19,6 +23,46 @@ public class PushToApplicationPreferences {
 
+
+    private PushToApplicationPreferences(){
+        this.activeApplicationName = new SimpleStringProperty("Texmaker");
```
<img width="945" height="71" alt="Image" src="https://github.com/user-attachments/assets/5e61dc62-787b-416f-89c0-4bbc45c068a2" />

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushToApplicationPreferences.java:
```diff
@@ -19,6 +23,46 @@ public class PushToApplicationPreferences {
+        commands.put("TeXstudio",  OS.detectProgramPath("texstudio", "TeXstudio"));
+        commands.put("TeXworks", OS.detectProgramPath("texworks", "TeXworks"));
+        commands.put("Sublime Text", OS.detectProgramPath("subl", "Sublime"));
+        commands.put("LyX/Kile", System.getProperty("user.home") + File.separator + ".lyx/lyxpipe");
```
```suggestion
        commands.put("LyX/Kile", USER_HOME + File.separator + ".lyx/lyxpipe");
```

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushToApplicationPreferences.java:
```diff
@@ -19,6 +23,46 @@ public class PushToApplicationPreferences {
+            commands.put("Emacs", "emacsclient.exe");
+        }else if(OS.OS_X || OS.LINUX){
+            commands.put("Emacs", "emacsclient");
+        }
```
```suggestion
        commands.put("Emacs", OS.WINDOWS ? "emacsclient.exe" : "emacsclient");
```

calixtus commented on jablib/src/main/java/org/jabref/logic/push/PushToApplicationPreferences.java:
```diff
@@ -19,6 +23,46 @@ public class PushToApplicationPreferences {
+
+    private PushToApplicationPreferences(){
+        this.activeApplicationName = new SimpleStringProperty("Texmaker");
+        Map<String, String> commands = new HashMap<>();
```
Why not FXCollections.observableMap here already?

koppor replied:
AI generated code

koppor replied:
AI generated code

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

calixtus commented:
Closing after 3 weeks unfinished draft
